### PR TITLE
*: support stale read

### PR DESCRIPTION
**Pull request by NingLin-P**

### CODE REVIEW

File: proto/kvrpcpb.proto [hunk 1]
```diff
@@ -769,6 +769,10 @@ message Context {
     // A hint for TiKV to schedule tasks more fairly. Query with same task ID
     // may share same priority and resource quota.
     uint64 task_id = 16;
+
+    // No required to read the most up-to-date data, replicas with `safe_ts` >= `start_ts`
```
Comment (BusyJay) [inline proto/kvrpcpb.proto:773]:
```suggestion
    // Not required to read the most up-to-date data, replicas with `safe_ts` >= `start_ts`
```

File: proto/raft_serverpb.proto [hunk 2]
```diff
@@ -23,6 +23,9 @@ message RaftMessage {
     // If it has value, to_peer should be removed if merge is never going to complete.
     metapb.Region merge_target = 9;
     ExtraMessage extra_msg = 10;
+    // use for stale read
+    uint64 applied_index = 11;
```
Comment (BusyJay) [inline proto/raft_serverpb.proto:27]: Why is it necessary?

File: proto/raft_serverpb.proto [hunk 3]
```diff
@@ -23,6 +23,9 @@ message RaftMessage {
...
+    uint64 safe_ts = 12;
```
Comment (BusyJay) [inline proto/raft_serverpb.proto:28]: I don't think this is something raft store can understand.

File: proto/raft_serverpb.proto (hunk 2 above)
Comment (NingLin-P) [reply proto/raft_serverpb.proto:27]: To forward follower's `safe_ts` through raft heartbeat

**Review by BusyJay — APPROVED** [review]

**Review by BusyJay — APPROVED** [review]

**Review by nolouch — APPROVED** [review]
lgtm

**Pull request merged**